topdown/eval.go

```diff
@@ -884,57 +903,122 @@ func (e *eval) Resolve(ref ast.Ref) (ast.Value, error) {
 		// example, a 2MB JSON value can take upwards of 30 millisceonds to convert.
 		// We cache the result of conversion here in case the same base document is
 		// being read multiple times during evaluation.
-		//
-		// NOTE(tsandall): If we introduce data mocking the cache will need to be
-		// invalidated.
-		if value := e.baseCache.Get(ref); value != nil {
-			e.instr.counterIncr(evalOpBaseCacheHit)
-			return value, nil
+
+		// check cache with mock data
+		mockValue, ok1 := e.withDataCache.Get(ref)
+		if ok1 {
+			return mockValue, nil
 		}
 
-		e.instr.counterIncr(evalOpBaseCacheMiss)
+		var merged ast.Value
+		var err error
 
-		path, err := storage.NewPathForRef(ref)
-		if err != nil {
-			if !storage.IsNotFound(err) {
-				return nil, err
+		// check cache with real data
+		realValue, ok2 := e.baseCache.Get(ref)
+		if ok2 {
+			e.instr.counterIncr(evalOpBaseCacheHit)
+
+			if mockValue == nil {
+				return realValue, nil
 			}
-			return nil, nil
-		}
 
-		blob, err := e.store.Read(e.ctx, e.txn, path)
-		if err != nil {
-			if !storage.IsNotFound(err) {
+			merged, err = merge(ref, mockValue, realValue)
+			if err != nil {
 				return nil, err
 			}
-			return nil, nil
-		}
+		} else {
+			e.instr.counterIncr(evalOpBaseCacheMiss)
+			path, err := storage.NewPathForRef(ref)
+			if err != nil {
+				if !storage.IsNotFound(err) {
+					return nil, err
+				}
+				return nil, nil
+			}
 
-		if len(path) == 0 {
-			obj := blob.(map[string]interface{})
-			if len(obj) > 0 {
-				cpy := make(map[string]interface{}, len(obj)-1)
-				for k, v := range obj {
-					if string(ast.SystemDocumentKey) == k {
-						continue
+			blob, err := e.store.Read(e.ctx, e.txn, path)
+			if err != nil {
+				if !storage.IsNotFound(err) {
+					return nil, err
+				}
+				return nil, nil
+			}
+
+			if len(path) == 0 {
+				obj := blob.(map[string]interface{})
+				if len(obj) > 0 {
+					cpy := make(map[string]interface{}, len(obj)-1)
+					for k, v := range obj {
+						if string(ast.SystemDocumentKey) == k {
+							continue
+						}
+						cpy[k] = v
 					}
-					cpy[k] = v
+					blob = cpy
 				}
-				blob = cpy
 			}
-		}
 
-		v, err := ast.InterfaceToValue(blob)
-		if err != nil {
-			return nil, err
+			v, err := ast.InterfaceToValue(blob)
+			if err != nil {
+				return nil, err
+			}
+
+			e.baseCache.Put(ref, v)
+
+			if mockValue == nil {
+				return v, nil
+			}
+
+			merged, err = merge(ref, mockValue, v)
+			if err != nil {
+				return nil, err
+			}
 		}
+		return merged, nil
+	}
+	return nil, fmt.Errorf("illegal ref")
+}
 
-		e.baseCache.Put(ref, v)
+func merge(ref ast.Ref, mockData, realData ast.Value) (ast.Value, error) {
+	_, ok1 := mockData.(ast.Object)
+	_, ok2 := realData.(ast.Object)
 
-		return v, nil
+	if ok1 && ok2 {
+		return mergeObjects(ref, mockData, realData)
 	}
+	return nil, mergeConflictErr(ref[0].Location)
```

Comment:
We'll only ever return an error if the two root values are not objects (since we only recurse on objects.) Is it impossible for there to be merge conflicts on non-root values?

Reply:
As long as the values being compared are objects we will recurse. If any of the values is not an object, we will overwrite the real value with the mock value.